### PR TITLE
Miscellaneous fixes and refactoring before task O4

### PR DESCRIPTION
**Pull request by cameel**
### Changes
This is a bunch of changes gathered from my work on task O4 which aren't really a part of its implementation. I put them all here because they don't introduce new functionality and would only make reviewing the actual solution harder. Some of these fix small issues in earlier branches or master. Others are just small improvements, cleanup or refactoring.

I recommend looking at individual commits if you really want to review it. Here's a simplified list of stuff that's changed:
1) Added a function that automatically builds a simple name for all cases tested with `parameterized()`. It's not perfect and names get weird when parameters are collections or complex strings but at least we get something more than a number without having to write a complex custom pattern for each test.
2) Refactored subtitle/data stream stripping functions and validations.
3) Fixed, cleaned up and added more information in the docstring for the `replace` command.
4) Cleaned up `.gitignore` and ignored `.mypy-cache`.
5) Added `can_convert()` for audio codecs which should have been implemented along with the one for video.
6) Rewritten, cleaned up and improved some tests.
    - Rewritten some bad tests for corner cases.
    - There were duplicate tests that were doing exactly the same thing.
    - I found one test for stream validation that was broken and not being running at all due to missing `test_` prefix.
    - I started using `mock.patch()` and `mock.patch.dict()` more to make unit tests more self-contained. This way we won't have to update the tests on every change to the dicts with codec/conversion definitions - only when the functionality itself changes. For example some tests rely on the fact that a specific codec/format/conversion isn't supported and making it supported can break those tests - sometimes even silently in such away that they pass but no longer test what they're supposed to.

        We'll also be able to tests situations that can't happen now but could if we changed the definitions (e.g. testing what happens when something is unsupported when all we have is currently marked as supported).

        This is only in several places right now but I'm using it more heavily in subsequent branches.
7) Tiny bugs in `Container`:
    - `list_supported_formats()` had a useless and unused `format` argument.
    - `is_supported_xxx_codec()` methods were silently ignoring unsupported types instead doing something to let you know that they don't support what you have given them.
8) Added lots of type annotations and fixed several incorrect ones.
9) Removed unused imports, stray commas, marked unused variables with `_`.

### API compatibility
Only refactoring and fixes. Zero changes in the API.

### Dependencies
This pull request is based on #24 and refactors code introduced in all the branches below so it can't be easily rebased directly on `master`.

**NOTE**: I have set the base branch of this pull request to `absolute-resource-paths-and-a-little-cleanup-in-tests` (#24). Please remember to change the base branch back to `master` before you merge.

### CODE REVIEW

File: tests/test_integration.py [hunk 1]
```diff
@@ -22,7 +22,7 @@ def setUp(self):
             'merge': os.path.join(self.tmp_dir, 'merge'),
             'replace': os.path.join(self.tmp_dir, 'replace'),
         }
-        for work_dir_id, work_dir_path in self.work_dirs.items():
+        for _work_dir_id, work_dir_path in self.work_dirs.items():
```
Comment (nieznanysprawiciel) [inline tests/test_integration.py]:
```suggestion
        for _, work_dir_path in self.work_dirs.items():
```

Comment (cameel) [reply tests/test_integration.py]: I always leave the name on purpose because it makes it easier to see what the function returns if I ever need it again. But it's just my personal preference. I changed it according to your suggestion.

Comment (cameel) [reply tests/test_integration.py]: This is already one by the way: c464ee24d0b598e26b348ad46bde33f62e21d3a8

**Comment by cameel** [comment]
> But make sure you have all necessary libraries in setup.py in tests_require list.

Sure. I looked at the code once more just in case and I don't think we're only using any external libraries beyond `parameterized` and `pytest`.
